[Infra UI] Add empty state content for waffle map #23260

XavierM opened the pull request:
@weltenwort @simianhacker 

Addresses (1) in https://github.com/elastic/kibana/issues/23260

I think if we need to add a button reload we should give back more information from the graphql services to make sure it is not an error. so we do not put the user in a loop of annoyance ;)

For now I did something really basic so the user know when it is loading and when there is no data.

Let me know if you want more than that

elasticcla commented:
Hi @XavierM, we have found your signature in our records, but it seems like you have signed with a different e-mail than the one used in yout Git [commit](https://github.com/elastic/kibana/pull/23319.patch). Can you please add both of these e-mails into your Github profile (they can be hidden), so we can match your e-mails to your Github profile?

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/2917/)

XavierM commented:
1- I added a spinner when you do a search with EuiLoadingChart component

![image](https://user-images.githubusercontent.com/189600/45771643-a5641400-bc13-11e8-9443-fc7692249e7a.png)

2 - Also added information if we are not bringing back any data from the search with EuiEmptyPrompt component

![image](https://user-images.githubusercontent.com/189600/45771736-e5c39200-bc13-11e8-8f99-3c3dfb2115f0.png)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/2945/)

tbragin commented:
@XavierM Nice! Thanks for working on this! I checked out the PR and took a look.

I noticed that "available" is misspelled, so we should definitely fix that. I would also go a step further and make this message visually consistent with what we have in the Logs UI, both in terms of fonts and icons. 

How about this?

<img width="883" alt="screen shot 2018-09-19 at 12 47 32 pm" src="https://user-images.githubusercontent.com/933397/45778279-148a3a00-bc0d-11e8-9f82-2924dd1a76ae.png">

We could also go a step further and adjust both messages to say something like this:

_Infra UI_

<b>There is no data to display.</b>
Try adjusting your time range or filter. 
Button: Check for new data

_Logs UI_

<b>There are no log messages to display.</b>
Try adjusting your time range or filter. 
Button: Check for new log messages

@weltenwort Thoughts?

weltenwort commented:
They look good to me, I like splitting them into a simple statement and the recommendation. As mentioned in #23260 I think we should avoid speaking about time ranges though. For waffle map the user only selects a point in time to sample for metrics. For the logging it's just a jump destination.

tbragin commented:
@weltenwort Make sense. Here is the updated proposal:

_Infra UI_

<b>There is no data to display.</b>
Try adjusting your time or filter. 
Button: Check for new data

_Logs UI_

<b>There are no log messages to display.</b>
Try adjusting your filter. 
Button: Check for new log messages

XavierM commented on x-pack/plugins/infra/public/components/waffle/index.tsx:
```diff
@@ -49,6 +50,24 @@ const calculateBoundsFromMap = (map: InfraWaffleData): InfraWaffleMapBounds => {
 export class Waffle extends React.Component<Props, {}> {
   public render() {
     const bounds = calculateBoundsFromMap(this.props.map);
+    if (this.props.loading) {
```
yes that make sense

XavierM commented on x-pack/plugins/infra/types/eui.d.ts:
```diff
@@ -72,7 +72,7 @@ declare module '@elastic/eui' {
 
   type EuiFilterButtonProps = CommonProps & {
-    color?: ButtonColor;
+    color?: IconColor;
```
You will have to show me where I can find the type in the eui web app

XavierM commented:
@tbragin @weltenwort 

To make the UI consistent I will advise that logs and infra both use the EmptyPrompt components out of the box from EUI https://elastic.github.io/eui/#/display/emptyprompt.

Does that make sense?

tbragin commented:
I'm good with the look and feel of it. It supports the structure we discussed. And I love the Dune reference, sold!

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3023/)

XavierM commented:
@tbragin I believe that we also want to match the loading screen as well.

Right now that's what we have for Logs, do you want the same for infra? and also should we say "Loading Logs"

![image](https://user-images.githubusercontent.com/189600/45837057-69e14c80-bcdc-11e8-9274-b24ef7536fcc.png)

tbragin commented:
Personally, I think "Loading entries" is fine, since these are log entries, but I would be okay with "Loading logs" as well. 

+1 on the same loading screen on the Infra UI and text saying "Loading data"

weltenwort commented:
Yes, there are several of those inconsistencies between the two parts of the UI from the time when they were separate apps. If you feel you want to align the page structures and use more modern EUI components, I think that would be great. :+1:

XavierM commented:
@tbragin 
Infra:
![image](https://user-images.githubusercontent.com/189600/45840776-f7756a00-bce5-11e8-8ece-781c4d0ca07b.png)
Logs:
![image](https://user-images.githubusercontent.com/189600/45840807-1116b180-bce6-11e8-807b-766928e56df3.png)

and loading are the same now

tbragin commented:
Looks good!

On the text for logging, per Felix, let's not mention time. Just "Try adjusting your filter".

XavierM commented:
Sorry, I missed that

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3076/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3087/)

tbragin commented:
LGTM

weltenwort commented on x-pack/plugins/infra/public/components/waffle/index.tsx:
```diff
@@ -48,11 +51,27 @@ const calculateBoundsFromMap = (map: InfraWaffleData): InfraWaffleMapBounds => {
+    } else if (!loading && map && map.length === 0) {
+      return (
+        <EuiEmptyPrompt
+          title={<h2>There is no data to display.</h2>}
```
Not being a native speaker I wonder whether it would be `There are no data [...]` since `data` is plural in itself? :thinking:

tbragin replied:
"There is no data" is grammatically correct in this context. http://www.onlinegrammar.com.au/top-10-grammar-myths-data-is-plural-so-must-take-a-plural-verb/

weltenwort replied:
Thanks, TIL about noncount nouns in english! :bulb:

XavierM commented on x-pack/plugins/infra/public/components/loading/index.tsx:
```diff
@@ -0,0 +1,47 @@
+  width: number | string;
+}
+
+export class InfraLoading extends React.PureComponent<InfraLoadingProps, {}> {
```
what about InfraLoadingScreen

weltenwort commented on x-pack/plugins/infra/public/containers/waffle/with_waffle_nodes.tsx:
```diff
@@ -49,12 +51,14 @@ export const WithWaffleNodes = ({
     }}
   >
-    {({ data }) =>
+    {({ data, loading, refetch }) =>
```
Without `notifyOnNetworkStatusChange` above, this will not rerender on `refetch`.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3159/)
